### PR TITLE
Migrate to JupyterLab 3

### PR DESCRIPTION
**Pull request by mlhenderson**
Updated to work with the current version of JupyterLab 3

Used the migration tool initially to update repo structure and package.json contents for compatibility (most new files in this PR come from this step)
Updated extension version in package.json to 3.0.0

Code modifications and cleanup:
 - Set classnames to constants in typescript with docstrings for future reference
 - Changed insert behavior for Favorites to be below the FileBrowser toolbar rather than above the BreadCrumbs (new FileBrowser file filtering component is properly grouped under FileBrowser now)
 - Updated CSS for BreadCrumbs Pinner Favorite icon to position it over the correct element
 - Moved contents of index.css to base.css, now index imports base

Remaining items

- [x] Update README with current install instructions for JupyterLab 3
- [x] Fix github build
- [x] Fix Travis CI build
- [x] Change source folder name to jupyterlab_favorites

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+include README.md
+include pyproject.toml
+include jupyter-config/jupyterlab-favorites.json
+
+include package.json
+include install.json
+include ts*.json
+include yarn.lock
+
+graft jupyterlab-favorites/labextension
```
Comment (fcollonval) [inline MANIFEST.in:11]:
```suggestion
graft jupyterlab_favorites/labextension
```

File: package.json [hunk 2]
```diff
@@ -16,48 +16,72 @@
...
-    "prepare": "npm run clean && npm run build",
-    "watch": "tsc -w"
+    "build": "jlpm run build:lib && jlpm run build:labextension:dev",
+    "build:labextension": "jupyter labextension build .",
+    "build:labextension:dev": "jupyter labextension build --development True .",
+    "build:lib": "tsc",
+    "build:prod": "jlpm run build:lib && jlpm run build:labextension",
+    "clean": "jlpm run clean:lib",
+    "clean:all": "jlpm run clean:lib && jlpm run clean:labextension",
+    "clean:labextension": "rimraf jupyterlab-favorites/labextension",
```
Comment (fcollonval) [inline package.json:41]:
```suggestion
    "clean:labextension": "rimraf jupyterlab_favorites/labextension",
```

File: package.json [hunk 3]
```diff
@@ -16,48 +16,72 @@
...
+    "npm-run-all": "^4.1.5",
+    "prettier": "^2.1.1",
+    "rimraf": "^3.0.2",
+    "typescript": "~4.1.3"
   },
   "jupyterlab": {
     "extension": true,
-    "schemaDir": "schema"
+    "schemaDir": "schema",
+    "outputDir": "jupyterlab-favorites/labextension"
```
Comment (fcollonval) [inline package.json:81]:
```suggestion
    "outputDir": "jupyterlab_favorites/labextension"
```

File: setup.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+else:
+    cmdclass["jsdeps"] = skip_if_exists(jstargets, js_command)
+
+long_description = (HERE / "README.md").read_text()
+
+# Get the package info from package.json
+pkg_json = json.loads((HERE / "package.json").read_bytes())
+
+setup_args = dict(
+    name=name,
```
Comment (fcollonval) [inline setup.py:61]:
```suggestion
    name=python_name,
```

File: setup.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    ensure_targets,
+    combine_commands,
+    skip_if_exists
+)
+import setuptools
+
+HERE = Path(__file__).parent.resolve()
+
+# The name of the project
+name = "jupyterlab-favorites"
```
Comment (fcollonval) [inline setup.py:19]:
```suggestion
name = "jupyterlab-favorites"
python_name = name.replace("-", "_")
```

File: setup.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+lab_path = (HERE / name / "labextension")
```
Comment (fcollonval) [inline setup.py:21]:
```suggestion
lab_path = (HERE / python_name / "labextension")
```

**Review by fcollonval — CHANGES_REQUESTED** [review]
@mlhenderson thanks a lot for working on this.

Small review from my side:
- Python package cannot contain `-` in their  name. Could you rename the python folder `jupyterlab_favorites` (+ some other places see my suggestions) ?
- Binary files make it into git. You should remove the files in the `dist` folder. I advice you to add that folder in the .gitignore file to avoid further trouble.

**Comment by mlhenderson** [comment]
@fcollonval Committing the binary files was intentional actually (see the PR comments), but I can definitely remove those if it is actually a problem, and then some editing of the README will also be needed.

**Comment by mlhenderson** [comment]
@fcollonval If this looks good to you now, let me know.

**Comment by fcollonval** [comment]
This looks good to me

> Committing the binary files was intentional actually (see the PR comments), but I can definitely remove those if it is actually a problem, and then some editing of the README will also be needed.

Sorry I miss that point. This is not recommended as binary files are not playing well with git and they by nature won't be in sync with the repository. Moreover their name will change at each release.

The recommended pattern is to publish them or to attach them to a GitHub release (you can drag & drop them on the release edition page). So the repository status corresponding to those artifacts is clear.

> Note: you can pip install a git repository - https://pip.pypa.io/en/stable/reference/pip_install/#git. If you want to mention that in the README.

**Comment by mlhenderson** [comment]
Good points, makes sense.  Thanks for the review!

**Comment by jasongrout** [comment]
Thank you *very* much!

**Pull request merged**